tests/testsuite/binary_name.rs

```diff
@@ -0,0 +1,352 @@
+[FINISHED] test [unoptimized + debuginfo] target(s) in [..]
+[RUNNING] [..] (target/debug/deps/foo-[..][EXE])",
+        )
+        .with_stdout_contains("test tests::check_crabs ... ok")
+        .run();
+
+    // Check if `cargo run` is able to execute the binary
+    p.cargo("run")
+        .masquerade_as_nightly_cargo()
+        .with_stdout("Hello, crabs!")
+        .run();
+
+    p.cargo("install").masquerade_as_nightly_cargo().run();
+
+    assert_has_installed_exe(cargo_home(), "007bar");
+
+    p.cargo("uninstall")
+        .with_stderr("[REMOVING] [ROOT]/home/.cargo/bin/007bar[EXE]")
+        .masquerade_as_nightly_cargo()
+        .run();
+
+    assert_has_not_installed_exe(cargo_home(), "007bar");
+}
+
+#[cargo_test]
+fn check_env_vars() {
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                cargo-features = ["different-binary-name"]
+
+                [project]
+                name =  "foo"
+                version = "0.0.1"
+
+                [[bin]]
+                name = "foo"
+                filename = "007bar"
+            "#,
+        )
+        .file(
+            "src/main.rs",
+            r#"
+                fn main() {
+                    println!("{}", option_env!("CARGO_BIN_NAME").unwrap());
+                }
+            "#,
+        )
+        .file(
+            "tests/integration.rs",
+            r#"
+                #[test]
+                fn check_env_vars2() {
+                    let value = option_env!("CARGO_BIN_EXE_$BIN_NAME").expect("Could not find environment variable.");
+                    assert!(value.contains("$BIN_NAME"));
+                }
+            "#
+            .replace("$BIN_NAME", "007bar")
+            .as_ref(),
+        )
+        .build();
+
+    // Run cargo build.
+    p.cargo("build").masquerade_as_nightly_cargo().run();
+    p.cargo("run")
+        .masquerade_as_nightly_cargo()
+        .with_stdout("007bar")
+        .run();
+    p.cargo("test")
+        .masquerade_as_nightly_cargo()
+        .with_status(0)
+        .run();
+}
+
+#[cargo_test]
+#[cfg(unix)]
+fn check_msg_format_json() {
+    // Create the project.
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                cargo-features = ["different-binary-name"]
+
+                [project]
+                name =  "foo"
+                version = "0.0.1"
+
+                [[bin]]
+                name = "foo"
+                filename = "007bar"
+                path = "src/main.rs"
+            "#,
+        )
+        .file("src/main.rs", "fn main() { assert!(true) }")
+        .build();
+
+    let exe_path = p.bin("007bar");
+
+    let output = |path| {
+        r#"
+{
+    "reason": "compiler-artifact",
+    "package_id": "foo 0.0.1 (path+file://[CWD])",
+    "manifest_path": "[CWD]/Cargo.toml",
+    "target": "{...}",
+    "profile": "{...}",
+    "features": [],
+    "filenames": "{...}",
+    "executable": "$EXE_PATH",
```

Comment:
This should make this test work on both unix and windows (and remove the closure and call to `replace`).  Also update `package_id` to use `[..]`.  The JSON comparison code uses the regular string comparison code that supports the matcher markup.

```suggestion
    "executable": "[ROOT]/foo/target/debug/007bar[EXE]",
```